packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -5,104 +5,132 @@
-
+import {AbortController} from "abort-controller";
 import {getTemporaryBlockHeader} from "@chainsafe/lodestar-beacon-state-transition";
 import {ILogger} from "@chainsafe/lodestar-utils";
 import {IBeaconDb} from "../../db";
-import {IEth1Notifier, Eth1Block, Eth1EventsBlock} from "../../eth1";
-import pipe from "it-pipe";
-import {IGenesisBuilder, IGenesisBuilderModules} from "./interface";
+import {
+  IDepositEvent,
+  IEth1StreamParams,
+  IEth1Provider,
+  getDepositsAndBlockStreamForGenesis,
+  getDepositsStream,
+} from "../../eth1";
+import {IGenesisBuilder, IGenesisBuilderModules, IGenesisResult} from "./interface";
 import {
   getGenesisBeaconState,
   getEmptyBlock,
   applyDeposits,
   applyTimestamp,
   applyEth1BlockHash,
   isValidGenesisState,
+  isValidGenesisValidators,
 } from "./util";
 
 export class GenesisBuilder implements IGenesisBuilder {
   private readonly config: IBeaconConfig;
   private readonly db: IBeaconDb;
-  private readonly eth1: IEth1Notifier;
+  private readonly eth1Provider: IEth1Provider;
+  private readonly eth1Params: IEth1StreamParams;
   private readonly logger: ILogger;
   private state: TreeBacked<BeaconState>;
   private depositTree: TreeBacked<List<Root>>;
 
-  constructor(config: IBeaconConfig, {eth1, db, logger}: IGenesisBuilderModules) {
+  private depositCache = new Set<number>();
+
+  constructor(config: IBeaconConfig, {db, eth1Provider, logger}: IGenesisBuilderModules) {
     this.state = getGenesisBeaconState(
       config,
       config.types.Eth1Data.defaultValue(),
       getTemporaryBlockHeader(config, getEmptyBlock())
     );
-    this.depositTree = config.types.DepositDataRootList.tree.defaultValue();
     this.config = config;
-    this.eth1 = eth1;
     this.db = db;
     this.logger = logger;
+    this.depositTree = config.types.DepositDataRootList.tree.defaultValue();
+    this.eth1Provider = eth1Provider;
+    this.eth1Params = {
+      ...config.params,
+      MAX_BLOCKS_PER_POLL: 10000,
+    };
   }
 
   /**
    * Get eth1 deposit events and blocks and apply to this.state until we found genesis.
    */
-  public async waitForGenesis(): Promise<TreeBacked<BeaconState>> {
-    await this.initialize();
-    const eth1DataStream = await this.eth1.getEth1BlockAndDepositEventsSource();
-    const state = await pipe(eth1DataStream, this.processDepositEvents(), this.assembleGenesisState());
-    await this.eth1.endEth1BlockAndDepositEventsSource();
-    return state;
-  }
+  public async waitForGenesis(): Promise<IGenesisResult> {
+    await this.eth1Provider.validateContract();
+
+    // TODO: Load data from data from this.db.depositData, this.db.depositDataRoot
+    // And start from a more recent fromBlock
+    const blockNumberLastestInDb = this.eth1Provider.deployBlock;
+
+    const blockNumberValidatorGenesis = await this.waitForGenesisValidators(blockNumberLastestInDb);
+
+    const controller = new AbortController();
```

Comment:
I think you should receive signal as method argument so high level module calling this can abort this upon "stop" method invocation